Add shortcut for Grisu3 algorithm.

mazong1123 opened the pull request:
While Grisu3 is way more faster for most numbers compare to Dragon4, the fall back to Dragon4 procedure for certain numbers could cause some performance regressions compare to use Dragon4 directly. Mitigating the regression caused by falling back is important for a largely used core library.

In Grisu3 algorithm implementation, there's a shortcut to jump out earlier when the fractional or integrals cannot meet the requirement of requested digits. This could significantly improve the performance of converting floating number to string as it falls back even without starting trying the algorithm.

The original idea is from the [.NET implementation](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Number.Grisu3.cs#L602-L615) and the code was originally added in [this PR](https://github.com/dotnet/coreclr/pull/14646#issuecomment-350942050). This shortcut has been shipped long time ago and has been proved working.

Fix #110129

mazong1123 commented:
@bors try @rust-timer queue

rust-timer commented:
Insufficient permissions to issue commands to rust-timer.
<!-- rust-timer: temporary -->

bors commented:
@mazong1123: :key: Insufficient privileges: not in try users

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4711747886/jobs/8356303180) [(plain)](https://github.com/rust-lang/rust/commit/766bbaa32b7653acf3f7ed35523b9e71ef4734df/checks/12778705658/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v3' (SHA:8e5e7e5ab8b370d6c329ec480221332ada57f0ab)
Download action repository 'rust-lang/simpleinfra@master' (SHA:75291a2cdf0dd5ba60860701274f9500703102da)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  CI_JOB_NAME: mingw-check-tidy
---
Building wheels for collected packages: reuse
  Building wheel for reuse (pyproject.toml): started
  Building wheel for reuse (pyproject.toml): finished with status 'done'
  Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=180119 sha256=9fa76c45f3193f307e02ea67d1a48cfe7ef2b854a074b766938a88e046bc7887
  Stored in directory: /tmp/pip-ephem-wheel-cache-ahz7kykc/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
Successfully built b7cb36fcd6b0
Successfully tagged rust-ci:latest
Built container sha256:b7cb36fcd6b0be20370777063abe385d4a3428a0a5adcda0b9e5efeda30b0564
Uploading finished image to https://ci-caches.rust-lang.org/docker/5a83b1174027be8472df586f36f1b4e77c39a20882cc84831c17b85c3cc5731b4e272c71c37dd18e09cacea39bf89c1b3226a304723983ec1270656c20627ec9
upload failed: - to s3://rust-lang-ci-sccache2/docker/5a83b1174027be8472df586f36f1b4e77c39a20882cc84831c17b85c3cc5731b4e272c71c37dd18e09cacea39bf89c1b3226a304723983ec1270656c20627ec9 Unable to locate credentials
[CI_JOB_NAME=mingw-check-tidy]
[CI_JOB_NAME=mingw-check-tidy]
---
   Compiling cargo_metadata v0.15.3
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished release [optimized] target(s) in 13.38s
fmt check
Diff in /checkout/library/core/src/num/flt2dec/strategy/grisu.rs at line 164:
 pub fn pow10_up_to_9(n: usize) -> u32 {
     debug_assert!(n <= 9);
 
-    const POW10: [u32; 10] = [
-        10,
-        100,
-        1000,
-        10_000,
-        10_000,
-        100_000,
-        1_000_000,
-        10_000_000,
-        100_000_000,
-        1_000_000_000,
-    ];
+    const POW10: [u32; 10] =
+        [1, 10, 100, 1000, 10_000, 100_000, 1_000_000, 10_000_000, 100_000_000, 1_000_000_000];
 
     POW10[n as usize]
Build completed unsuccessfully in 0:00:14
Build completed unsuccessfully in 0:00:14
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/library/test/src/cli.rs" "/checkout/library/core/src/num/flt2dec/mod.rs" "/checkout/library/core/src/num/flt2dec/estimator.rs" "/checkout/library/core/src/num/flt2dec/strategy/dragon.rs" "/checkout/library/core/src/num/flt2dec/strategy/grisu.rs" "/checkout/library/core/src/num/flt2dec/decoder.rs" "/checkout/library/core/src/num/bignum.rs" "/checkout/library/core/src/num/mod.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

mazong1123 commented:
@simulacrum looks like I don't have permission to request a perf benchmark. As this change is perf related, is there any chance to grant me the right permission?

mazong1123 commented:
Here is the [benchmark data](https://github.com/dotnet/coreclr/pull/14646#issuecomment-351593345) for .NET with the shortcut. Doing similar benchmark tests in rust would be great.

fee1-dead commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 9968365655939cad46b5ec70ff3f1d201d150ba8 with merge 09a6ce7b9a3eb5f779b7015c5138e9b8c54ff00c...
<!-- homu: {"type":"TryBuildStarted","head_sha":"9968365655939cad46b5ec70ff3f1d201d150ba8","merge_sha":"09a6ce7b9a3eb5f779b7015c5138e9b8c54ff00c"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4711895134/jobs/8357171105)
Build commit: 09a6ce7b9a3eb5f779b7015c5138e9b8c54ff00c (`09a6ce7b9a3eb5f779b7015c5138e9b8c54ff00c`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4711895134/jobs/8357171105"},"merge_sha":"09a6ce7b9a3eb5f779b7015c5138e9b8c54ff00c"} -->

rust-timer commented:
Queued 09a6ce7b9a3eb5f779b7015c5138e9b8c54ff00c with parent 2a711152615ad9294dc0e5ee6885c8e9bb8418a9, future [comparison URL](https://perf.rust-lang.org/compare.html?start=2a711152615ad9294dc0e5ee6885c8e9bb8418a9&end=09a6ce7b9a3eb5f779b7015c5138e9b8c54ff00c).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (09a6ce7b9a3eb5f779b7015c5138e9b8c54ff00c): [comparison URL](https://perf.rust-lang.org/compare.html?start=2a711152615ad9294dc0e5ee6885c8e9bb8418a9&end=09a6ce7b9a3eb5f779b7015c5138e9b8c54ff00c&stat=instructions:u).

### Overall result: no relevant changes - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=2a711152615ad9294dc0e5ee6885c8e9bb8418a9&end=09a6ce7b9a3eb5f779b7015c5138e9b8c54ff00c&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=2a711152615ad9294dc0e5ee6885c8e9bb8418a9&end=09a6ce7b9a3eb5f779b7015c5138e9b8c54ff00c&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 2.0%  | [2.0%, 2.0%]   | 1     |
| Improvements ✅ <br /> (primary)   | -3.7% | [-3.7%, -3.7%] | 1     |
| Improvements ✅ <br /> (secondary) | -3.9% | [-5.2%, -2.4%] | 3     |
| All ❌✅ (primary)                 | -3.7% | [-3.7%, -3.7%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=2a711152615ad9294dc0e5ee6885c8e9bb8418a9&end=09a6ce7b9a3eb5f779b7015c5138e9b8c54ff00c&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 4.7% | [4.3%, 5.0%] | 3     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

fee1-dead commented:
This can be expected since rustc perf is mostly related to the performance of the compiler. You should consider writing your own benchmarks to see the performance changes.

mazong1123 commented:
Appears current rust benchmark for Grisu3 does not have numbers to cover the fallback cases. I've added 3 benchmarks: `bench_one_exact_inf`, `bench_trailing_zero_exact_inf` and `bench_halfway_point_exact_inf`. Below are the benchmark results in my local machine:

# Before change
benchmarks:
    num::flt2dec::strategy::grisu::bench_big_exact_12             75.00ns/iter  +/- 4.00ns
    num::flt2dec::strategy::grisu::bench_big_exact_3              35.00ns/iter  +/- 1.00ns
    num::flt2dec::strategy::grisu::bench_big_exact_inf            33.15µs/iter  +/- 5.16µs
    num::flt2dec::strategy::grisu::bench_big_shortest             93.00ns/iter  +/- 5.00ns
    **num::flt2dec::strategy::grisu::bench_halfway_point_exact_inf 475.00ns/iter +/- 25.00ns**
    **num::flt2dec::strategy::grisu::bench_one_exact_inf           476.00ns/iter +/- 20.00ns**
    num::flt2dec::strategy::grisu::bench_small_exact_12           58.00ns/iter  +/- 3.00ns
    num::flt2dec::strategy::grisu::bench_small_exact_3            32.00ns/iter  +/- 1.00ns
    num::flt2dec::strategy::grisu::bench_small_exact_inf           1.00µs/iter +/- 27.00ns
    num::flt2dec::strategy::grisu::bench_small_shortest           51.00ns/iter  +/- 2.00ns
    **num::flt2dec::strategy::grisu::bench_trailing_zero_exact_inf 490.00ns/iter +/- 16.00ns**

# After change
benchmarks:
    num::flt2dec::strategy::grisu::bench_big_exact_12             77.00ns/iter   +/- 3.00ns
    num::flt2dec::strategy::grisu::bench_big_exact_3              35.00ns/iter   +/- 1.00ns
    num::flt2dec::strategy::grisu::bench_big_exact_inf            33.11µs/iter   +/- 2.61µs
    num::flt2dec::strategy::grisu::bench_big_shortest             92.00ns/iter   +/- 7.00ns
    **num::flt2dec::strategy::grisu::bench_halfway_point_exact_inf 413.00ns/iter  +/- 31.00ns**
    **num::flt2dec::strategy::grisu::bench_one_exact_inf           416.00ns/iter  +/- 65.00ns**
    num::flt2dec::strategy::grisu::bench_small_exact_12           62.00ns/iter   +/- 3.00ns
    num::flt2dec::strategy::grisu::bench_small_exact_3            32.00ns/iter   +/- 1.00ns
    num::flt2dec::strategy::grisu::bench_small_exact_inf         985.00ns/iter +/- 319.00ns
    num::flt2dec::strategy::grisu::bench_small_shortest           54.00ns/iter   +/- 8.00ns
    **num::flt2dec::strategy::grisu::bench_trailing_zero_exact_inf 434.00ns/iter  +/- 26.00ns**

| Benchmark  | Before | After | Improvement |
| ------------- | -------------  | -------------  | ------------- |
| (num::flt2dec::strategy::grisu::bench_halfway_point_exact_inf)  | 475.00ns/iter | 413.00ns/iter | 13.05% |
| (num::flt2dec::strategy::grisu::bench_one_exact_inf)  | 476.00ns/iter |  416.00ns/iter | 12.60%|
| num::flt2dec::strategy::grisu::bench_trailing_zero_exact_inf   | 490.00ns/iter | 434.00ns/iter | 11.42% |

mazong1123 commented:
@rustbot review

mazong1123 commented:
@Mark-Simulacrum if you have time :)

Mark-Simulacrum commented:
Please be aware the reviews can take 7-14 days; the reviewer pool is predominantly composed of volunteers. I will get to this fairly soon, likely today or tomorrow, but rest assured that as long as it's in S-waiting-on-review it's in my queue.

Mark-Simulacrum commented on library/core/src/num/flt2dec/strategy/grisu.rs:
```diff
@@ -159,6 +159,17 @@ pub fn max_pow10_no_more_than(x: u32) -> (u8, u32) {
+/// Given `n`, where `0 <= n <= 9`, returns `10^n`.
+#[doc(hidden)]
+pub fn pow10_up_to_9(n: usize) -> u32 {
+    debug_assert!(n <= 9);
```
This can be dropped -- indexing the array gives you a hard assert that n is in `[0, 9)`.

I might just move the array into the function below and index directly as well, seems simpler. This function shouldn't need to be public regardless.

Mark-Simulacrum commented:
r=me with nit fixed

mazong1123 committed:
Add shortcut for Grisu3 algorithm.

Check requested digit length and the fractional or integral parts of the number. Falls back earlier without trying the Grisu algorithm if the specific condition meets.

Fix #110129

mazong1123 commented:
Thanks @Mark-Simulacrum . Updated accordingly. Commits are squashed as well.

fee1-dead commented:
@bors r=Mark-Simulacrum

bors commented:
:pushpin: Commit b0a85d614d5cc1918bc9eef9abbf78d5c585cbef has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum b0a85d614d5cc1918bc9eef9abbf78d5c585cbef -->
<!-- homu: {"type":"Approved","sha":"b0a85d614d5cc1918bc9eef9abbf78d5c585cbef","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit b0a85d614d5cc1918bc9eef9abbf78d5c585cbef with merge 91b61a4ad618c1abc2af43a58695de185ef1e513...
<!-- homu: {"type":"BuildStarted","head_sha":"b0a85d614d5cc1918bc9eef9abbf78d5c585cbef","merge_sha":"91b61a4ad618c1abc2af43a58695de185ef1e513"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4794644387/jobs/8530818028)
Approved by: Mark-Simulacrum
Pushing 91b61a4ad618c1abc2af43a58695de185ef1e513 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Mark-Simulacrum","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4794644387/jobs/8530818028"},"merge_sha":"91b61a4ad618c1abc2af43a58695de185ef1e513"} -->

rust-timer commented:
Finished benchmarking commit (91b61a4ad618c1abc2af43a58695de185ef1e513): [comparison URL](https://perf.rust-lang.org/compare.html?start=ed3ddb375bf18099a885605efba94dc1e50379d7&end=91b61a4ad618c1abc2af43a58695de185ef1e513&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=ed3ddb375bf18099a885605efba94dc1e50379d7&end=91b61a4ad618c1abc2af43a58695de185ef1e513&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=ed3ddb375bf18099a885605efba94dc1e50379d7&end=91b61a4ad618c1abc2af43a58695de185ef1e513&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 2.6% | [2.6%, 2.6%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=ed3ddb375bf18099a885605efba94dc1e50379d7&end=91b61a4ad618c1abc2af43a58695de185ef1e513&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.
